src/asset_tracking.rs

```diff
@@ -0,0 +1,57 @@
+use bevy::prelude::*;
+use std::collections::VecDeque;
+
+pub(super) fn plugin(app: &mut App) {
+    app.init_resource::<ResourceHandles>()
+        .add_systems(PreUpdate, load_resource_assets);
+}
+
+pub trait LoadResource {
+    /// This will load the [`Resource`] as an [`Asset`]. When all of its asset dependencies
+    /// have been loaded, it will be inserted as a resource. This ensures that the resource only
+    /// exists when the assets are ready.
+    fn load_resource<T: Resource + Asset + Clone + FromWorld>(&mut self) -> &mut Self;
+}
+
+impl LoadResource for App {
+    fn load_resource<T: Resource + Asset + Clone + FromWorld>(&mut self) -> &mut Self {
+        self.init_asset::<T>();
+        let world = self.world_mut();
+        let value = T::from_world(world);
+        let assets = world.resource::<AssetServer>();
+        let handle = assets.add(value);
+        let mut handles = world.resource_mut::<ResourceHandles>();
+        handles
+            .waiting
+            .push_front((handle.untyped(), |world, handle| {
+                let assets = world.resource::<Assets<T>>();
+                if let Some(value) = assets.get(handle.id().typed::<T>()) {
+                    world.insert_resource(value.clone());
+                }
+            }));
+        self
+    }
+}
+
+#[derive(Resource, Default)]
+struct ResourceHandles {
+    waiting: VecDeque<(UntypedHandle, fn(&mut World, &UntypedHandle))>,
+    #[allow(unused)]
+    finished: Vec<UntypedHandle>,
+}
+
+fn load_resource_assets(world: &mut World) {
+    world.resource_scope(|world, mut resource_handles: Mut<ResourceHandles>| {
+        world.resource_scope(|world, assets: Mut<AssetServer>| {
+            for _ in 0..resource_handles.waiting.len() {
```

Comment:
```suggestion
            for _ in &resource_handles.waiting {
```

Reply:
Not the same thing, as it holds a borrow to `resource_handles.waiting`, but we modify resource_handles.waiting in the loop.

Reply:
You're right. Changed my comment while you were typing :D

Reply:
>  could we not simplify this by making it a vec

I don't see how that is any simpler. We'd still be doing the same operations. Might as well preserve the order.

Reply:
While the code would look the same, I think the mental burden would be smaller. The simplicity comes from not having to consider the two-tailedness of operations. IME many people consider `Vec` to be the "default" container, considering its simplicity it being the only one with a convenience macro and a list or vector often being the first data structure that programmers learn to use in a new language. As such, a usage of `VecDeque` IMO sticks out as "Using this was a very important and well-thought out decision, please pay attention to it", which is not really the case. The order being preserved here is fairly arbitrary imo, and in a world where we already merged a version using `Vec` and someone opened a PR for a `VecDeque`, I'd probably reject that PR as unnecessary.
If you feel strongly about this, we can let it be.

Reply:
Reasonable. I don't feel strongly. Pushed the change!